### PR TITLE
[SourceKit] Add option to index local symbols to IndexSource request

### LINKED ISSUE #69077: [SourceKit] Add option to index local symbols to IndexSource request
Currently, the `source.request.indexsource` response includes only global symbols. Within the codebase, the locals indexing is mostly supported and even used in the `swift-ide-test` test utility. This functionality could be exposed to the clients by adding optional `key.should_index_locals` boolean flag into the `source.request.indexsource` request.

In case this is a desirable improvement, I already implemented this in a PR:  #69076  🙂

### PR DESCRIPTION
**Pull request by barinsim**
Adds optional ~~`key.should_index_locals`~~ `key.include_locals` boolean field into the ~~`source.request.indexsource`~~ `source.request.index_to_store`  request. If this field is set to `true`, the response includes local symbols alongside the global symbols.

Resolves #69077

### CODE REVIEW

**Comment by bnbarham** [comment]
Couple initial thoughts:
  1. It would be better to use the new `source.request.index_to_store` request rather than this one (https://github.com/apple/swift/pull/66958). I'd ideally like to remove this in the future.
  2. Couldn't you just include `-index-include-locals` in the arguments?

**Comment by barinsim** [comment]
@bnbarham 

Sorry for a late reply 🙂 

> It would be better to use the new source.request.index_to_store request rather than this one

Ok, I did not know that the `indexsource` is on its way out. I moved the option into the `index_to_store` request.

> Couldn't you just include -index-include-locals in the arguments?

Currently adding the flag into the `key.compilerargs` does not seem to work. The flag gets dropped while we are building `CompilerInvocation` to generate the AST. I did not dig much deeper.

But I think that this flag is not much different from the `key.index_store_path` which also has the corresponding `-index-store-path` compiler argument. Was it ever discussed what kinds of compiler arguments should have an explicit SourceKit request field and which should not? For example `-sdk` looks natural in the  `key.compilerargs` but I don't know where to draw the line.

**Comment by bnbarham** [comment]
> Currently adding the flag into the key.compilerargs does not seem to work. The flag gets dropped while we are building CompilerInvocation to generate the AST. I did not dig much deeper.

In the old request or with `index_to_store`? We're explicitly passing `Opts.IndexIncludeLocals` so I would expect it to work for the new request and if not consider it a bug.

> But I think that this flag is not much different from the key.index_store_path which also has the corresponding -index-store-path compiler argument. Was it ever discussed what kinds of compiler arguments should have an explicit SourceKit request field and which should not? For example -sdk looks natural in the  key.compilerargs but I don't know where to draw the line.

Honestly I'm unclear why I didn't bring this up in the review at the time. @DavidGoldman do you remember why we went with arguments to the request vs compiler arguments here? Looking at it again, I would prefer passing both the store path and unit output path as compiler arguments + pulling out `emitIndexDataForSourceFile` (used during compilation) and using it for the request as well.

The main argument I can think of is that we would need to add these to *all* requests, otherwise the args would be different and then we'd have to rebuild the AST.

**Comment by barinsim** [comment]
@bnbarham Hi, I just wanted to check the status on this. Do you think it can be merged as it is right now (after CI run)? i.e. the option is part of `index_to_store` and must be provided in the explicit `include_locals` field.

I am for the explicit field option because this

>The main argument I can think of is that we would need to add these to all requests, otherwise the args would be different and then we'd have to rebuild the AST.

looks like a good reason to me.

>In the old request or with index_to_store? We're explicitly passing Opts.IndexIncludeLocals so I would expect it to work for the new request and if not consider it a bug.

I mean the `index_to_store`. But with the explicit option it is no longer a problem as the compiler option gets dropped during `SwiftInvocationRef Invok = ASTMgr->getTypecheckInvocation(Args, PrimaryFilePath, Error)` but the later `emitIndexDataForSourceFile()` work correctly.

cc: @DavidGoldman

File: tools/SourceKit/lib/SwiftLang/SwiftIndexing.cpp [hunk 1]
```diff
@@ -393,7 +394,7 @@ static void emitIndexDataForSourceFile(SourceFile &PrimarySourceFile,
                                !Opts.IndexIgnoreClangModules,
                                Opts.IndexSystemModules,
                                Opts.IndexIgnoreStdlib,
-                               Opts.IndexIncludeLocals,
+                               IncludeLocals,
```
Comment (bnbarham) [inline tools/SourceKit/lib/SwiftLang/SwiftIndexing.cpp:397]: Could you pull out the rest of these as well?

File: test/SourceKit/Indexing/index_locals.swift [hunk 2]
```diff
@@ -0,0 +1,86 @@
+// RUN: %empty-directory(%t)
+// RUN: %sourcekitd-test -req=index-to-store %s -index-store-path %t/idx -index-unit-output-path %t/index_locals.o -req-opts=include_locals=1 -- %s
+// RUN: c-index-test core -print-record %t/idx | %FileCheck %s
```
Comment (bnbarham) [inline test/SourceKit/Indexing/index_locals.swift:3]: IMO it's enough to just test that a single local is indexed here, assuming we have a test for locals themselves in test/Index. If we don't then I'd prefer most of that going there and then a simple test here to make sure the option itself works.

**Review by bnbarham — COMMENTED** [review]
Just have a couple other comments. I'm happy enough to have the options, though I do think that the compiler arguments should at the very least be the defaults. It doesn't make sense to ignore them completely. Fixing that doesn't have to be in this PR though.

Out of interest, what are you using the index data for (if you can say)?

**Comment by skrtks** [comment]
Hi @bnbarham, Simon completed an internship with us at JetBrains. He's currently finishing his thesis before rejoining us on a full-time basis. I don't want to bother him too much in the meantime, so I implemented the changes you requested and asked Simon to merge them into this pr.

> Out of interest, what are you using the index data for (if you can say)?

We are using the index as part of our efforts to phase out our resolve engine for Swift. With the rapid evolution and increasing complexity of the language, it was too much work for our small team to maintain it.

**Comment by bnbarham** [comment]
> so I implemented the changes you requested and asked Simon to merge them into this pr.

Awesome, thanks 🙇‍♂️ 

> We are using the index as part of our efforts to phase out our resolve engine for Swift. With the rapid evolution and increasing complexity of the language, it was too much work for our small team to maintain it.

Oh, interesting. Thanks for the info!

**Comment by bnbarham** [comment]
@swift-ci please test

**Review by bnbarham — APPROVED** [review]
Thanks @barinsim and @skrtks!

**Pull request merged**